iwyu.cc

```diff
@@ -2224,6 +2224,32 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     return true;
   }
 
+  bool VisitTypeTraitExpr(clang::TypeTraitExpr* expr) {
+    if (CanIgnoreCurrentASTNode())  return true;
+
+    // We assume that all type traits with >= 2 arguments require
+    // full type information even for pointer types. For example,
+    // this is the case for `__is_convertible_to` trait.
+    if (expr == nullptr || expr->getNumArgs() <= 1)  return true;
+
+    for (const clang::TypeSourceInfo* arg : expr->getArgs()) {
+      auto qual_type = arg->getType();
+      const Type* type = qual_type.getTypePtr();
+      if (const PointerType* pointer_type =
+              dyn_cast<PointerType>(type)) {
+        const Type* deref_type = pointer_type->getPointeeType().getTypePtr();
+        ReportTypeUse(CurrentLoc(), deref_type);
+      } else {
+        // This isn't really required (underlying types will be
+        // traversed anyway), by makes type traits handling logic
```

Comment:
typo: by -> but

I'm not sure about the "more self-contained" bit. It sounds like you don't need to do anything extra anywhere for the else-leg, right, it's just redundant reporting?

See also `VisitBinaryOperator`, https://github.com/include-what-you-use/include-what-you-use/blob/master/iwyu.cc#L2214. It seems to me like the arguments here could be `T**` in which case `getPointeeType()` will return `T*`? 

You could add a testcase for double indirection (and C++ references!) and see what happens. It could be that you need to strip off all pointers and references here to get to the underlying type. There should be a support function for that in `iwyu_ast_util.h`.

Also the `CanIgnoreType` call might be worth adding. I haven't fully understood what that does, but it seems like if we _can_ ignore a type, we probably should.

Reply:
Thanks for the suggestion about `iwyu_ast_util.h`. I ended up using `RemovePointersAndReferencesAsWritten` function. Also I added `CanIgnoreType` check.